Team417_CENTERSTAGE/src/main/java/org/firstinspires/ftc/team417_CENTERSTAGE/mechanisms/ArmTeleOp.java

```diff
@@ -0,0 +1,249 @@
+    //positionsArray: array of positions in encoder ticks, currentPosition: location of mechanism in encoder ticks from init location
+    //acceptableError: the amount of difference between the current location and a position in the array where it will still consider itself past the position
+    private double findLastPosition(double[] positionsArray, double currentPosition, double acceptableError) {
+        double lastPosition = 0;
+
+        for (int i = positionsArray.length - 1; i >= 0; i--) {
+            if (positionsArray[i] < currentPosition - acceptableError) {
+                lastPosition = positionsArray[i];
+                break;
+            }
+        }
+
+        return lastPosition;
+    }
+
+    public void armControl() {
+        //How much moving the right stick will effect arm speed outside and inside the robot.
+        double rStickSensitivity = 1, rStickSensitivityInsideRobot = rStickSensitivity / 3, rStickSensitivityOutsideRobot = rStickSensitivity;
+
+        double[] ArmPositions = new double[] {BaseOpMode.ARM_MOTOR_MIN_POSITION, BaseOpMode.ARM_MOTOR_MAX_POSITION / 2.0,
+                //array of arm positions the dpad can move to.
+                BaseOpMode.ARM_MOTOR_MAX_POSITION * (3.0/4.0), BaseOpMode.ARM_MOTOR_MAX_POSITION};
+
+        //The current distance the arm is from its init location
+        double armCurrentLocation = armMotor.getCurrentPosition();
+
+        //The amount the right stick has been moved
+        double rStickTilt = -gamepad2.right_stick_y;
+
+        //Defines a 0.1 deadzone for right stick input; allows a 25-unit error for arm position goals.
+        final double rStickDeadZone = 0.1, armPosEpsilon = 25;
+
+        //Locations the dumper should tilt so the arm does not hit the robot.
+        final double startUpwardTiltPos = 75, startUpwardResetPos = 1500, endUpwardResetPos = 2000, startDownwardTiltPos = 1500, endDownwardTiltPos = 250;
+
+        //The speed the arm moves while in move to position mode
+        final double armMoveToPositionVelocity = 0.5;
+
+        //If the arm is outside the dead zone switch to using stick control
+        if (gamepad2.left_stick_y < -rStickDeadZone || gamepad2.left_stick_y > rStickDeadZone) {
+            usingStick = true;
+
+        //If dpad up pressed, move arm to next position in array with armPosEpsilon tolerance.
+        } else if (gamepad2.dpad_up && !dpadUpPressed) {
+            armGoalLocation = findNextPosition(ArmPositions, armCurrentLocation, armPosEpsilon);
+            usingStick = false;
+
+        //If dpad down pressed, move arm to last position in array with armPosEpsilon tolerance.
+        } else if (gamepad2.dpad_down && !dpadDownPressed) {
+            armGoalLocation = findLastPosition(ArmPositions, armCurrentLocation, armPosEpsilon);
+            usingStick = false;
+        }
+
+        //If arm is inside robot decrease the speed so arm can't hit robot.
+        if (armMotor.getCurrentPosition() < 150.0)
```

Comment:
This should also be a constant. Please address all magic numbers throughout your code including ones I have not marked.